### PR TITLE
Add control of volume scaling in Barostat

### PR DESCRIPTION
**Pull request by badisa**
* Get and set the volume scaling
* Disable/enable adaptive scaling of the volume scaling

### CODE REVIEW

File: timemachine/cpp/src/barostat.hpp [hunk 1]
```diff
@@ -30,11 +31,21 @@ template <typename RealType> class MonteCarloBarostat {
 
     int get_interval();
 
+    double get_volume_scaling();
+
+    void set_volume_scaling(const double scaling);
```
Comment (mcwitt) [inline timemachine/cpp/src/barostat.hpp:36]: nit: naming: maybe `{get,set}_volume_scale_factor` to better align with the terminology [here](http://docs.openmm.org/latest/userguide/theory/02_standard_forces.html#montecarlobarostat)?

File: timemachine/cpp/src/barostat.hpp [hunk 2]
```diff
@@ -30,11 +31,21 @@ template <typename RealType> class MonteCarloBarostat {
...
     void set_pressure(const double pressure);
 
+    void set_adaptive_scaling(const bool adapative);
+
+    bool get_adaptive_scaling();
+
 private:
     const int N_;
 
+    bool adaptive_; // Whether or no to adapt d_volume_scaling
```
Comment (mcwitt) [inline timemachine/cpp/src/barostat.hpp:47]:
```suggestion
    bool adaptive_; // Whether or no to adapt d_volume_scale_
```

File: timemachine/cpp/src/kernels/k_barostat.cuh [hunk 3]
```diff
@@ -128,7 +130,7 @@ void __global__ k_decide_move(
     const bool rejected = w > 0 && rand[1] > std::exp(-w / kt);
 
     while (idx < N) {
-        if (idx == 0) {
+        if (idx == 0 && adaptive) {
```
Comment (mcwitt) [inline timemachine/cpp/src/kernels/k_barostat.cuh:133]: nit: this seems like it also disables accumulation of accept/reject counts when `adaptive` is false? Is there a benefit to maintaining the stats? (e.g. would they be useful for diagnostics, or to better inform moves the next time `adaptive` is enabled)

File: timemachine/cpp/src/wrap_kernels.cpp [hunk 4]
```diff
@@ -1202,23 +1202,30 @@ void declare_barostat(py::module &m) {
...
             py::arg("temperature"),
             py::arg("group_idxs"),
             py::arg("frequency"),
             py::arg("bps"),
-            py::arg("seed"))
+            py::arg("seed"),
+            py::arg("adaptive") = true)
         .def("set_interval", &Class::set_interval, py::arg("interval"))
         .def("get_interval", &Class::get_interval)
+        .def("set_volume_scaling", &Class::set_volume_scaling, py::arg("volume_scaling"))
```
Comment (mcwitt) [inline timemachine/cpp/src/wrap_kernels.cpp:1225]: nit: consider `py::arg("volume_scale_factor")` if suggestion to change the getter and setter names is implemented

File: timemachine/cpp/src/kernels/k_barostat.cuh (hunk 3 above)
Comment (maxentile) [reply timemachine/cpp/src/kernels/k_barostat.cuh:133]: +1 for maintaining acceptance/trial counters even when not currently being used to adapt proposal scale.

File: tests/test_barostat.py [hunk 5]
```diff
@@ -545,3 +545,89 @@ def test_get_group_indices():
...
+    # Initial volume scaling is 0
+    assert baro.get_volume_scaling() == 0.0
+    assert baro.get_adaptive_scaling()
+
+    ctxt = custom_ops.Context(coords, v_0, host_box, integrator.impl(), u_impls, barostat=baro)
+    ctxt.multiple_steps(15)
+
+    # Verify that the volume scaling is non-zero
+    scaling = baro.get_volume_scaling()
+    assert scaling >= 0
```
Comment (mcwitt) [inline tests/test_barostat.py:612]:
```suggestion
    assert scaling > 0
```
to match comment

File: timemachine/cpp/src/wrap_kernels.cpp [hunk 6]
```diff
@@ -1202,23 +1202,30 @@ void declare_barostat(py::module &m) {
...
-                   std::vector<std::shared_ptr<timemachine::BoundPotential>> bps,
-                   const int seed) { return new Class(N, pressure, temperature, group_idxs, frequency, bps, seed); }),
+            py::init([](const int N,
+                        const double pressure,
+                        const double temperature,
+                        std::vector<std::vector<int>> group_idxs,
+                        const int frequency,
+                        std::vector<std::shared_ptr<timemachine::BoundPotential>> bps,
+                        const int seed,
+                        const bool adaptive) {
```
Comment (maxentile) [inline timemachine/cpp/src/wrap_kernels.cpp:1212]: nit[naming]: slight preference for more descriptive `adapt_proposal_scale` or similar https://github.com/proteneer/timemachine/blob/61e068297ce26083b1168535ba908f9987696605/timemachine/md/barostat/moves.py#L92

File: timemachine/cpp/src/kernels/k_barostat.cuh [hunk 7]
```diff
@@ -99,12 +100,13 @@ void __global__ k_setup_barostat_move(
...
 void __global__ k_decide_move(
     const int N,
+    const bool adaptive,
     const int num_molecules,
     const double kt,
     const double pressure,
     const RealType *__restrict__ rand, // [2] Use second value
     RealType *__restrict__ d_volume_delta,
-    RealType *__restrict__ d_volume_scale,
+    double *__restrict__ d_volume_scale,
```
Comment (maxentile) [inline timemachine/cpp/src/kernels/k_barostat.cuh:109]: nit[type surprise]: should `d_volume_delta` and `d_volume_scale` share the same type?

File: tests/test_barostat.py [hunk 8]
```diff
@@ -545,3 +545,89 @@ def test_get_group_indices():
...
+        coords.shape[0],
+        pressure,
+        temperature,
+        group_indices,
+        barostat_interval,
+        u_impls,
+        seed,
+    )
+    # Initial volume scaling is 0
+    assert baro.get_volume_scaling() == 0.0
```
Comment (maxentile) [inline tests/test_barostat.py]: nit[non-corner case as default]: preference for this to be a non-zero value.

(I believe OpenMM's MonteCarloBarostat initializes with something like `proposal_scale = 0.01 * volume(initial_box)` or similar.)

**Review by maxentile — APPROVED** [review]
Looks good!

(Some superficial nits, but this feature will be helpful in ruling out one tricky-to-reason-about source of possible sampling bias when the same stateful barostat instance is used by multiple chains.)

File: tests/test_barostat.py (hunk 8 above)
Comment (badisa) [reply tests/test_barostat.py]: It does get initialized to that value, but when you construct the barostat it doesn't know what the volume is. 

https://github.com/proteneer/timemachine/blob/master/timemachine/cpp/src/kernels/k_barostat.cuh#L87-L89

File: timemachine/cpp/src/kernels/k_barostat.cuh (hunk 7 above)
Comment (badisa) [reply timemachine/cpp/src/kernels/k_barostat.cuh:109]: They don't have to, will switch d_volume_delta back

File: timemachine/cpp/src/wrap_kernels.cpp (hunk 6 above)
Comment (badisa) [reply timemachine/cpp/src/wrap_kernels.cpp:1212]: Would `adapt_volume_scale_factor` be better here or should `volume_scale_factor -> proposal_scale_factor`?

**Review by mcwitt — APPROVED** [review]

File: tests/test_barostat.py (hunk 8 above)
Comment (maxentile) [reply tests/test_barostat.py]: Ahh, thanks for clarifying!

(Less confident: but may be useful for this to always be positive, even at initialization (maybe 1% of `volume(4A x 4A 4A solvent box)`?))

**Review by proteneer — APPROVED** [review]

File: timemachine/cpp/src/wrap_kernels.cpp (hunk 6 above)
Comment (maxentile) [reply timemachine/cpp/src/wrap_kernels.cpp:1212]: >Would adapt_volume_scale_factor be better here

No preference

>should volume_scale_factor -> proposal_scale_factor?

Current name `volume_scale_factor` is less ambiguous 👍

File: timemachine/cpp/src/kernels/k_barostat.cuh (hunk 7 above)
Comment (badisa) [reply timemachine/cpp/src/kernels/k_barostat.cuh:109]: Cchanged back in [2ebdfa5](https://github.com/proteneer/timemachine/pull/1152/commits/2ebdfa5b6309fefc61e599a00ebfe05af7af99a7)

File: timemachine/cpp/src/wrap_kernels.cpp (hunk 6 above)
Comment (badisa) [reply timemachine/cpp/src/wrap_kernels.cpp]: Changed in [c08bd5c](https://github.com/proteneer/timemachine/pull/1152/commits/c08bd5c3e2f21532582f2ccdfdc7d97e86598edf)

File: timemachine/cpp/src/wrap_kernels.cpp (hunk 4 above)
Comment (badisa) [reply timemachine/cpp/src/wrap_kernels.cpp]: Done in [5c39c88](https://github.com/proteneer/timemachine/pull/1152/commits/5c39c8893119ec43434328d333b41ae638a5cf05)

File: timemachine/cpp/src/barostat.hpp (hunk 1 above)
Comment (badisa) [reply timemachine/cpp/src/barostat.hpp]: Changed in [5c39c88](https://github.com/proteneer/timemachine/pull/1152/commits/5c39c8893119ec43434328d333b41ae638a5cf05)

File: timemachine/cpp/src/kernels/k_barostat.cuh (hunk 3 above)
Comment (badisa) [reply timemachine/cpp/src/kernels/k_barostat.cuh]: Changed to maintain counter in [7d71499](https://github.com/proteneer/timemachine/pull/1152/commits/7d71499b1b972c6c8f45f6b2726b4901acdd84d0)

File: tests/test_barostat.py (hunk 8 above)
Comment (badisa) [reply tests/test_barostat.py]: What sort of case are you imagining this to apply? If it was some default value, it could be significantly off and would probably need to be re-initialized and you would want to reinitialize. For example, if it was set for a 4A box, and you gave it a 10A box it seems like it could take significantly longer to reach equilibrium.

Comment (maxentile) [reply tests/test_barostat.py]: Sorry, I was just confused.

(Wanted to avoid the case where a proposal scale of 0.0 was ever used. Reading more carefully, although the proposal scale can _nominally_ be 0.0, it will never _effectively_ be 0.0. Enforcing that this value is nominally always positive -- while retaining the logic "the first time you see an input configuration, reset proposal scale to 1% of its volume" -- wouldn't actually do anything different than the current implementation.)

File: timemachine/cpp/src/barostat.hpp [hunk 9]
```diff
@@ -30,11 +31,21 @@ template <typename RealType> class MonteCarloBarostat {
 
     int get_interval();
 
+    double get_volume_scale_factor();
+
+    void set_volume_scale_factor(const double scaling);
+
     void set_pressure(const double pressure);
 
+    void set_adaptive_scaling(const bool adapative);
```
Comment (mcwitt) [inline timemachine/cpp/src/barostat.hpp:40]: nit: typo
```suggestion
    void set_adaptive_scaling(const bool adaptive);
```

File: timemachine/cpp/src/barostat.hpp [hunk 10]
```diff
@@ -30,11 +31,21 @@ template <typename RealType> class MonteCarloBarostat {
 
     int get_interval();
 
+    double get_volume_scale_factor();
+
+    void set_volume_scale_factor(const double scaling);
```
Comment (mcwitt) [inline timemachine/cpp/src/barostat.hpp:36]: nit:
```suggestion
    void set_volume_scale_factor(const double scale_factor);
```

File: timemachine/cpp/src/barostat.hpp (hunk 9 above)
Comment (mcwitt) [reply timemachine/cpp/src/barostat.hpp]: nit: prefer more descriptive argument name, maybe `adaptive_scaling_enabled`?

File: timemachine/cpp/src/barostat.hpp [hunk 11]
```diff
@@ -19,7 +19,8 @@ template <typename RealType> class MonteCarloBarostat {
         std::vector<std::vector<int>> group_idxs,
         const int interval,
         std::vector<std::shared_ptr<BoundPotential>> bps,
-        const int seed);
+        const int seed,
+        const bool adapt_volume_scale_factor);
```
Comment (mcwitt) [inline timemachine/cpp/src/barostat.hpp:23]: nit: generally it seems like we should be consistent between getter/setter names, names of the setter argument, and the name of the private variable that's being updated

Comment (badisa) [reply timemachine/cpp/src/barostat.hpp:23]: Unified in [b102405](https://github.com/proteneer/timemachine/pull/1152/commits/b10240596853c2ab57047e8b153a6536fe373412)

File: timemachine/cpp/src/barostat.hpp (hunk 10 above)
Comment (badisa) [reply timemachine/cpp/src/barostat.hpp]: Fixed in [b102405](https://github.com/proteneer/timemachine/pull/1152/commits/b10240596853c2ab57047e8b153a6536fe373412)
